### ISSUE TITLE
Disable "Apply" button upon applying

### ISSUE THREAD

**Issue by vedxyz** [OpenTabletDriver/OpenTabletDriver#2123]
*From the UI design discussion on Discord*

## Description

Disable the "Apply" button when the settings have just been applied and no new changes exist.

Also, right now, the apply button staying enabled implies that the settings have to be "Save"d as well.
This would make it a bit more obvious what the difference between "Save" and "Apply" is, and that "Save" is not needed in every situation.

**Comment by X9VoiD** [comment]
What's more is that "Save" is also "Apply" in that it applies what is changed, and then saves it. So people that usually pressed apply then save is actually applying their settings *twice*.

It can be further improved by not applying on save, then disabling "Save" when changes are not applied yet, also disable save when already saved.

**Comment by vedattt** [comment]
> by not applying on save, then disabling "Save" when changes are not applied yet, also disable save when already saved.

I can see how this would be nice to make things completely explicit, especially if the disabled buttons have concise tooltips to show why they are disabled.

On the other hand though, I think "Save" also functioning as an *apply then save* is more conventional. For instance, settings windows usually have an "OK" button that usually works as a *apply then close*, which I find similar to our use case.

I'd say I like it both ways, but it's some minor food for thought.

**Comment by Sublimelime** [comment]
>On the other hand though, I think "Save" also functioning as an apply then save is more conventional. For instance, settings windows usually have an "OK" button that usually works as a apply then close, which I find similar to our use case.

I agree with this. We want to keep save as an apply+save function no matter what we end up doing. You'd never want to save settings without applying them too.

The only thing that really needs to be done here is to just indicate in some way that settings have been applied. We could disable the apply button, pop up some kind of temporary header that says "Applied!" or something, etc. With the latter, people could learn that save also applies, so they'd stop double-applying.

**Comment by vedattt** [comment]
> You'd never want to save settings without applying them too.

Void's suggestion does cover this, since you wouldn't be able to "Save" without first "Apply"ing. So the workflow would be:

1. Save [disabled] | Apply [enabled] -- *user has changed a setting*
2. Save [enabled] | Apply [disabled] -- *user has clicked apply*
3. Save [disabled] | Apply [disabled] -- *user has clicked save*
4. Save [disabled] | Apply [enabled] -- *user has changed a setting, back to step 1*

You might call this a hassle, and you wouldn't exactly be wrong lol.
I guess lengthening the process complicates things unnecessarily, so I'm in favor of

- Save => apply then save (always enabled, or maybe disabled upon save)
- Apply => apply (then disable until further change)

**Comment by vedattt** [comment]
Actually, as a final option, I suggest to take Void's idea but modify it such that "Save" is still _apply then save_.
That gives us these states:

*Workflow 1*
1. Save [enabled] | Apply [enabled] -- *user has changed a setting*
2. Save [enabled] | Apply [disabled] -- *user has clicked **apply***
3. Save [disabled] | Apply [disabled] -- *user has clicked **save***

*Workflow 2*
1. Save [enabled] | Apply [enabled] -- *user has changed a setting*
2. Save [disabled] | Apply [disabled] -- *user has clicked **save***

Same as Void's idea, this has 3 distinct states, but the user experience is more sensible this way imo.
It should be the implementation's responsibility to avoid a double apply, and preventing it is trivial in this case as you would simply check if "Apply" is disabled when "Save"ing.

I'm strongly in favor of this last option.

**Comment by X9VoiD** [comment]
+1 to this last suggestion.

**Comment by gonX** [comment]
Milestone set to v0.6.6 since this is fixed in Avalonia.

**Comment by gonX** [comment]
I'm cleaning up the v0.6.7 milestone - and I've been wondering if this is even something that's worth putting time into for fixing in 0.6 at all?

I'm not super familiar with this part of Eto but I feel like we would have to wire up every single input to an "invalidation bool" that could be used to tell if the Apply button needs disabling.

If there's an interesting way to fix this with Eto I'd gladly take it, but the naive method I proposed seems prone to bugs and involves a couple of months of "oh wow" pull requests to iron out any eventual bugs that that implementation might have.

Just trying to be realistic - I do personally want to stop adding new features for 0.6 that can't be translated in some way to the new GUI, and `avalonia` branch has already accounted for this as mentioned in my prior message.

Thoughts @Mrcubix @Kuuuube ?

TL;DR I want to close as wontfix for 0.6 and as resolved for Avalonia?

**Comment by gonX** [comment]
After playing around a bit with MVVM, if our GUI infrastructure is any good we should be able to simply connect the `Enabled` state of the button to the settings object. The settings object might however need to implement `INotifyPropertyChanged` if it doesn't already.

**Comment by Kuuuube** [comment]
One suggestion that might be worth looking at is to simply disable the button for a set amount of time (maybe 1-2 seconds) after clicking on it. This way users are given feedback and can be confident they clicked the button and there should be less hassle with the code.

A full implementation can be left for later.

**Comment by gonX** [comment]
This doesn't seem easy to implement in 0.6. We would need to wire up a lot of places where settings can change. I'm gonna close this as 'wontfix' for 0.6.x and fixed in `avalonia`. #4458 mitigates this issue in 0.6.
